lib/Parse/ParseDecl.cpp

```diff
@@ -5769,24 +5769,30 @@ Parser::parseDeclEnumCase(ParseDeclOptions Flags,
 
       // For recovery, see if the user typed something resembling a switch
       // "case" label.
-      llvm::SaveAndRestore<decltype(InVarOrLetPattern)>
-      T(InVarOrLetPattern, Parser::IVOLP_InMatchingPattern);
-      parseMatchingPattern(/*isExprBasic*/false);
-
-      if (consumeIf(tok::colon)) {
-        diagnose(CaseLoc, diag::case_outside_of_switch, "case");
-        Status.setIsParseError();
-        return Status;
-      }
-      if (CommaLoc.isValid()) {
-        diagnose(Tok, diag::expected_identifier_after_case_comma);
-        Status.setIsParseError();
-        return Status;
+      {
+        BacktrackingScope backtrack(*this);
+        llvm::SaveAndRestore<decltype(InVarOrLetPattern)>
+        T(InVarOrLetPattern, Parser::IVOLP_InMatchingPattern);
+        parseMatchingPattern(/*isExprBasic*/false);
+        
+        if (consumeIf(tok::colon)) {
+          backtrack.cancelBacktrack();
+          diagnose(CaseLoc, diag::case_outside_of_switch, "case");
+          Status.setIsParseError();
+          return Status;
+        }
       }
+      
       if (NameIsKeyword) {
         diagnose(TokLoc, diag::keyword_cant_be_identifier, TokText);
         diagnose(TokLoc, diag::backticks_to_escape)
           .fixItReplace(TokLoc, "`" + TokText.str() + "`");
+        if (!Tok.isAtStartOfLine())
+          consumeToken();
```

Comment:
Could you update `Name` and `NameLoc` with that token one? (That might improve diagnostics in later stage)

Reply:
@rintaro Update the `NameLoc` is fine, but update name location causes an `invalid redeclaration of '_'` diagnostic when `case Snout, _;` or `  case _`. We could only update `Name` when `!Tok.isIdentifierOrUnderscore()` or just `Tok.isNot(tok::kw_`) (since is inside NameIsKeyword) but just update the NameLoc seems fine to me, what do you think?

Reply:
`invalid redeclaration of '_'` is OK because there *is* redeclaration.
```swift
enum E {
 case Snout, _;
 case _ // invalid redeclaration of '_'.
}
```
The second `_` should be treated as redeclaration of `_`.
Just to make sure, there is no redeclaration error if there's only one `_` in the enum decl, right?
```swift
enum E {
  case Snout
  case _
}
```

Reply:
Yeah, I think there isn't on the second case (have to confirm) and is indeed a redeclaration. But, since _ (without backtricks) is not a valid identifier, does it makes sense to trigger a invalid redeclaration, because the declaration itself is not valid?

Reply:
@rintaro, I've updated that, but if you think the question above is valid I can just revert the last commit :))

Reply:
Ah thanks. I think redeclaration error should be emitted even if it's caused by unescaped keyword like `_`.